From b87df900f5af2e99d64ef80b72b2ca01f6a475f2 Mon Sep 17 00:00:00 2001
From: Assam Boudjelthia <assam.boudjelthia@qt.io>
Date: Wed, 10 Dec 2025 21:45:03 +0200
Subject: [PATCH 25/25] Android: don't enable focus for QtEditText when added
 to the view

This way we avoid having the QtInputConnection invoke any operations
that might end up locking the the UI thread before the potential
window surfaces are being created on the UI thread and end up not
being able to acquire it.

Pick-to: 6.8 6.5
Fixes: QTBUG-141579
Fixes: QTBUG-141782
Fixes: QTBUG-141357
Fixes: JTP-23
Fixes: JTP-4
Fixes: JTP-3
Task-number: QTBUG-139400
Task-number: QTBUG-142460
Change-Id: I152f0dc3976b28e7ba589958d3470f9c0d3dcccb
Reviewed-by: Ville Voutilainen <ville.voutilainen@qt.io>
Reviewed-by: Jani Korteniemi <jani.korteniemi@qt.io>
(cherry picked from commit e1be7cce120be42cc91083b37bf95f9082f04106)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
(cherry picked from commit 3e06f1010ff81eea7f094478ca8738c514179b73)
---
 .../jar/src/org/qtproject/qt/android/QtWindow.java    | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/src/android/jar/src/org/qtproject/qt/android/QtWindow.java b/src/android/jar/src/org/qtproject/qt/android/QtWindow.java
index f0d304f9e7f..c63cd1a77e3 100644
--- a/src/android/jar/src/org/qtproject/qt/android/QtWindow.java
+++ b/src/android/jar/src/org/qtproject/qt/android/QtWindow.java
@@ -32,6 +32,7 @@ class QtWindow extends QtLayout implements QtSurfaceInterface {
     private QtWindow m_parentWindow;
     private GestureDetector m_gestureDetector;
     private final QtEditText m_editText;
+    private boolean m_editTextFocusInitialized = false;
     private final QtInputConnection.QtInputConnectionListener m_inputConnectionListener;
     private boolean m_firstSafeMarginsDelivered = false;
     private int m_actionBarHeight = -1;
@@ -62,6 +63,8 @@ class QtWindow extends QtLayout implements QtSurfaceInterface {
         if (!isForeignWindow && context instanceof Activity) {
             // TODO QTBUG-122552 - Service keyboard input not implemented
             m_editText = new QtEditText(context, listener);
+            m_editText.setFocusable(false);
+            m_editText.setFocusableInTouchMode(false);
             m_editText.setImportantForAccessibility(View.IMPORTANT_FOR_ACCESSIBILITY_NO);
             LayoutParams layoutParams = new LayoutParams(ViewGroup.LayoutParams.WRAP_CONTENT,
                     ViewGroup.LayoutParams.WRAP_CONTENT);
@@ -242,6 +245,14 @@ class QtWindow extends QtLayout implements QtSurfaceInterface {
     @Override
     public boolean onTouchEvent(MotionEvent event)
     {
+        // Enable focus for the edit text on first touch event to avoid
+        // early QtInputConnection callbacks from blocking the UI thread.
+        if (!m_editTextFocusInitialized) {
+            m_editTextFocusInitialized = true;
+            m_editText.setFocusable(true);
+            m_editText.setFocusableInTouchMode(true);
+        }
+
         windowFocusChanged(true, getId());
         if (m_editText != null && m_inputConnectionListener != null)
             m_inputConnectionListener.onEditTextChanged(m_editText);
-- 
2.52.0

